contributors/design-proposals/bulk_watch.md

```diff
@@ -0,0 +1,459 @@
+```
+ We can't simply make this an http GET request, due to limitations of GET for
+the size (length) of the url (in which we would have to pass filter options).
+
+ We could consider adding `watch` operation using the same pattern with just
+`?watch=1` parameter. However, the main drawback of this approach is that it
+won't allow for dynamic altering of watch subscriptions (which we definitely
+also need to support).
+As a result, we need another API for watch that will also support incremental
+subscriptions - it will look as following:
+```
+websocket /apis/bulk.k8s.io/v1/list?watch=1
+```
+
+*Note: For consistency, we also considered introducing websocket API for
+handling LIST requests, where first client sends a filter definition over the
+channel and then server sends back the response, but we dropped this for now.*
+
+*Note: We aso considered implementing the POST-based watch handler that doesn't
+allow for altering subsriptions, which should be very simple once we have list
+implemented. But since websocket API is needed anyway, we also dropped it.*
+
+
+### Filtering definition
+
+ We need to make our filtering mechanism to support different resource types at
+the same time. On the other hand, we would like it to be as consistent with all
+other Kubernetes APIs as possible. So we define the selector for bulk operations
+as following:
+
+```
+type BulkListOptions struct {
+	Selectors []ListSelector
+}
+
+type ListSelector struct {
+	Resource GroupVersionResource
+
+	// We would like to reuse the same ListOptions definition as we are using
+	// in regular APIs. For the purpose of bulk operations, you can just specify
+	// more than one.
+	//
+	// NOTE: In the initial implementation, we will require all Options (for a
+	// given resource type) to have exactly the same ResourceVersion - otherwise
+	// we will reject it at the validation level.
+	// This will simplify the initial implementation as we need to guarantee that
+	// watch events within a single resource types are delivered with increasing
+	// resource versions, which may be more tricky with different resource versions.
+	Options []ListOptions
+}
+```
+
+ We will create a dedicated admission plugin (or other filter mechanism) that
+will be responsible for detecting whether a given user is allowed to list/watch
+objects requested by a given `BulkdSelector` (the exact mechanism for it is out
+of scope for this document) and either rejecting the whole request or letting it
+go. We are not going to support partial rejections (e.g. you cannot proceed with
+first ListSelector in the list, but you can with the second one). For example,
+Kubelet will be allowed to list/watch only objects that are "attached" to pods
+bound to a given node.
+
+ As mentioned in the Proposal section, we will also be periodically verifying
+ACLs for watches and breaking the watch whenever ACLs change and user is no
+longer allowed to watch all requested objects.
+
+
```

Comment:
@lavalamp - sorry, when I moved the "POST http"-based watch section to the bottom, it removed your marger.

It was here :)